src/prefect/agent/local/agent.py

```diff
@@ -72,7 +72,7 @@ def __init__(
         )
 
     def heartbeat(self) -> None:
-        for idx, process in enumerate(self.processes):
+        for idx, process in enumerate(list(self.processes)):
```

Comment:
This is still incorrect, but in a more subtle and dangerous way. Before this would error since you can't mutate a collection over which you're iterating. This now succeeds, but will pop out incorrect elements, since the length of the list changes with each mutation.

- Loop over list with 3 elements `[a, b, c]`
- `idx, process = 0, a`, pop out idx 0, list is now `[b, c]`
- `idx, process = 1, b`, pop out idx 1, list is now `[b]` - *the wrong element was removed*

I think `.processes` should be a set of processes instead of a list, and use `self.processes.remove(process)` to remove completed processes. This provides O(1) remove for individual processes (more efficient than the `O(n)` call to `pop` here), and doesn't run the risk of removing the wrong process.

Reply:
Oops, you're right. In my own internal conception I was just putting the remaining processes into a new list and swapping that one in place.

Reply:
PR incoming soon.

Reply:
Oh damn @trkohler I let you astray 🤦

Reply:
it's nice to learn! especially, when it comes with such brilliant explanations